docs/docs/configuration.md
---
layout: default
title: Configuration
nav_order: 3
---

# Configuration
{: .no_toc }


deemon has some specific configuration parameters that can be defined in your 
config.json file.
{: .fs-6 .fw-300 }

## Table of contents
{: .no_toc .text-delta }

1. TOC
{:toc}

---

## Configuration Location
Depending on your operating system, your config.json file will be located in 
one of the locations below. When deemon is run with a command, it will 
automatically generate a default config if an existing configuration file is 
not present. For example: to generate this configuration file, run 
`deemon refresh`.

- **Linux**: /home/user/.config/deemon

- **macOS**: /User/user/Library/Application Support/deemon

- **Windows**: %appdata%\deemon

### Default Configuration (Version 2.19.2)
```json
{
    "check_update": 1,
    "debug_mode": false,
    "release_channel": "stable",
    "query_limit": 5,
    "smart_search": true,
    "rollback_view_limit": 10,
    "prompt_duplicates": false,
    "prompt_no_matches": true,
    "fast_api": true,
    "fast_api_threads": 25,
    "exclusions": {
        "enable_exclusions": true,
        "patterns": [],
        "keywords": [],
    },
    "new_releases": {
        "release_max_age": 90,
        "include_unofficial": false,
        "include_compilations": false,
        "include_featured_in": false,
    },
    "global": {
        "bitrate": "320",
        "alerts": false,
        "record_type": "all",
        "download_path": "",
        "email": ""
    },
    "deemix": {
        "path": "",
        "arl": "",
        "check_account_status": true,
        "halt_download_on_error": false,
    },
    "smtp_settings": {
        "server": "",
        "port": 465,
        "starttls": false,
        "username": "",
        "password": "",
        "from_addr": ""
    },
    "plex": {
        "base_url": "",
        "ssl_verify": true,
        "token": "",
        "library": ""
    }
}
```

## Configuration Types

There are technically three different "levels" of configuration in deemon. The 
first level (and default) is the "global" configuration, the second level is 
the "profile" configuration and the third level is the "per-artist" 
configuration.

### Global Configuration

The global configuration is defined inside the `config.json` configuration 
file. This configuration is used by default when running deemon and the values 
defined in this configuration may be temporarily superseded by a 
[Profile](#profile-configuration-optional) or an 
[Artist](#per-artist-configuration-optional) configuration.

### Profile Configuration *(optional)*

Within deemon, you can configure multiple profiles (using the `profile` 
command) which monitors their own sets of artists and can use settings that 
override the _Global Configuration_ while inheriting the settings that are not 
defined. You can also think of these profiles as separate users. For example, 
if you create a new profile and only specify a download path, that profile will 
inherit the other settings from the _Global Configuration_.

### Per-Artist Configuration *(optional)*

Within deemon, you can configure different settings for each individual artist 
(using the `config` command). This gives you the flexibility to disable alerts 
for certain artists or specify a certain release type, bitrate or download path.

<br>

## Global Configuration
This section outlines each setting available in the configuration file and their respective options.

---

### Global settings
These settings affect the general operation of deemon.

|Setting|Description|
|-|---|
|**check_update**<br><br><br>|This option allows you to specify how frequently (in days) to check for new updates to deemon. To disable checking for updates, change this to `0`.<br><br>|
|**debug_mode**<br>options: _true, false_<br><br>|This option will allow you to print extra debug messages in the logs or on screen if used with `--verbose`.<br><br>|
|**release_channel**<br>options: _stable, beta_<br><br><br>|When checking for updates (if enabled), this option allows you to choose what updates you are notified about. Most users will want to use _stable_. If you are interested in testing pre-release versions of deemon, you can set this to _beta_.<br><br>|
|**query_limit**<br>options: _number_<br><br><br>|This option allows you to specify how many results are displayed when using the `search` command or when prompted using the `monitor` command (see _prompt_duplicates_ and _prompt_no_matches_ below).<br><br>|
|**smart_search**<br>options: _true, false_<br><br><br>|This option allows you to skip the list of artist search results and proceed directly to the list of artist albums, provided there is only one exact match of the artists name (case insensitive).<br><br><br>|
|**rollback_view_limit**<br><br><br>|This option allows you to specify the maximum number of transactions to display using the `rollback` command<br><br>|
|**prompt_duplicates**<br>options: _true, false_<br><br><br><br>|When adding a new artist using the `monitor` command, deemon will choose the highest ranked artist in situations where two artists have identical names. Instead, you can set this option to `true` which will prompt you with choices including the latest release from each artist to help you better decide which is the artist you're looking for.<br><br>|
|**prompt_no_matches**<br>options: _true, false_<br><br><br>|When adding a new artist using the `monitor` command, if deemon does not find an **exact** match for the artist you're searching for, it will prompt you with a list of results returned from the Deezer API.<br><br>|
|**fast_api**<br>options: _true, false_<br><br>|In previous versions of deemon, this was referred to as the _experimental_api_ and has been the default API since version 2.1.<br><br>|
|**fast_api_threads**<br>options: _number_<br><br>|This sets the number of threads to spawn when accessing the API. The higher the number, the faster artist data is retrieved. However, setting this number too high may result in a temporary ban of your IP address. **It is recommended to keep this number below 50.**<br><br>|

---

### Exclusion settings
Exclusions can be setup to ignore releases matching a specific regular expression ("pattern") or by matching against specific keywords. You can test your exclusion settings by using the `test` command.

|Setting|Description|
|-|---|
|**enable_exclusions**<br>options: _true, false_<br><br>|This setting enables exclusion patterns and keywords (below) to filter out releases.<br><br><br>|
|**patterns**<br>options: _regex_<br><br>|Provide a list of regular expressions to filter out releases.<br><br><br>|
|**keywords**<br>options: _true, false_<br><br>|Provide a list of keywords to filter out releases (_remix, deluxe, bonus_).<br><br><br>|

---

### New Release settings
These settings affect what releases are filtered out by deemon. As of version 2.9, these settings are global only and cannot be configured in a profile nor per-artist configuration.

|Setting|Description|
|-|---|
|**release_max_age**<br><br><br><br><br><br><br><br>|This lets you control how far back to consider a new release 'new' from it's actual release date. This setting is helpful if you only run a `refresh` monthly or if there is a delay in Deezer adding a release to their catalogue. If you wish to get all releases regardless of when they were released, set this to _0_.<br><br>**Note:** Version 2.8.x and earlier relied on a separate toggle _by_release_date_ which has been deprecated in favor of setting _release_max_age_ to _0_.<br><br>|
|**include_unofficial**<br>options: _true, false_<br><br><br><br><br><br>|Each release result returned from the API includes a boolean value designating the release as either an official artist release or not. In many cases, this flag is incorrectly set resulting in some official releases not being detected.<br><br>**Warning:** If you are enabling this option on an existing database, you could potentially have hundreds or thousands of new releases detected.<br><br>|
|**include_compilations**<br>options: _true, false_<br><br><br><br><br>|If you want to include all compilation albums that your artists are featured on, you can enable this.<br><br>**Warning:** If you are enabling this option on an existing database, you could potentially have hundreds or thousands of new releases detected.<br><br>|
|**include_featured_in**<br>options: _true, false_<br><br><br><br><br><br>|Enabling this option will include every single release/track an artist is associated with. It is highly recommend to avoid using this but is included for testing purposes.<br><br>**Warning:** If you are enabling this option on an existing database, you could potentially have **_tens of thousands_** of new releases detected. Most users will NOT want this option enabled.<br><br>|

---

### Global settings
These settings can be overriden within deemon using _profiles_ or by specifying a _per-artist configuration_.

|Setting|Description|
|-|---|
|**bitrate**<br>options: _128, 320, FLAC_<br><br>|This option allows you to specify the bitrate used for downloads.<br><br><br>|
|**alerts**<br>options: _true, false_<br><br>|Enable or disable email notification alerts when new releases are present. You must also have your email settings configured (see _[SMTP Settings](#smtp-settings)_).<br><br>|
|**record_type**<br>options: all, album, ep, single<br><br><br>|This option allows you to specify what release types you wish to download upon release. Keep in mind, most EPs are labelled as albums in the Deezer API.<br><br>**Limitation:** Currently only one option at a time is allowed.<br><br>|
|**download_path**<br><br><br><br><br><br>|This option allows you to specify where downloads are saved. If no path is provided, downloads will be saved in the default directory specified by deemix.<br><br>**Windows users:** When providing a path, you **must** use double backslashes: `C:\\Music` or forward slashes: `C:/Music`.<br><br>|
|**email**<br><br><br>|This option allows you to specify the default email address to use when alerts are enabled and SMTP settings are defined.<br><br>|

---

### deemix settings
These settings are needed for deemon to interface with deemix which is a third party library that does the actual downloading.

|Setting|Description|
|-|---|
|**path**<br><br><br>|Specify the path to where your deemix configuration is stored. For most users, leave this blank. If you have moved the deemix configuration, you must specify that here.<br><br>|
|**arl**<br><br><br><br>|This is your authorization token required by `deemix` to authenticate your Deezer account. This is stored in a cookie named `arl` in your browser after logging in to Deezer.<br><br>|

CHANGELOG
Version 2.0                                    Sept. 24, 2021
=============================================================
Added:
  Highlights:
    - Multi-profile support
    - Interactive search/download client
    - Artist Configuration tool
    - Restore backups right from inside deemon
    - Smarter monitoring ensuring more accurate results
    - New config system with config validation
    - Rollback monitor/refresh transactions
  General:
    - Added email notifications on expired ARLs
  Backup
    - Restore backups using new backup manager `backup --restore`
  Config (command)
    - Change per-artist configuration using new `config` command
    - To remove a particular value in artist config, type "none"
  Configuration
    - Added option to switch between stable/beta update notifications
    - Specify `query_limit` limits the amount of search results for `monitor` and `search`
    - Set number of days between update checks with `check_update`
    - New settings for _monitor_ command: `prompt_no_matches` and `ranked_duplicates`
  Download
    - Download all releases from a given date forward using `download --from-date`
  Monitor
    - Prompt user to choose if no exact match is found
    - Duplicate artist names are now auto-selected based on rank
    - Specify a download path on a per-artist basis
  Profile
    - New `profile` command for managing profiles and their configurations
    - Configuration profiles for monitoring a separate collection of artists
    - Specify profile to use at run-time: `deemon -P NAME ...`
  Refresh
    - Refresh artists individually
    - Progress bars are now dynamic and fit the screen
    - Simulate refresh using new `--dry-run` option
  Rollback
    - New `rollback` command to undo monitor/refresh transactions
    - Specify limit of transactions to see in `rollback --view` using `rollback_view_limit`
  Search
    - `search` command for searching and downloading new music
    - Ability to monitor/stop monitoring an artist from artist menu
    - Queue multiple artists, albums and tracks and download at once
    - Filter albums by type and sort by year/title
    - Explicit album toggle to change visibility of such albums
  Show
    - Export even more data with new CSV filtering options
    - Export artist data to CSV file
    - Playlists now visible using `show playlists`
    - When using `--csv`, default view is to show all fields
    - Apply specified filters to CSV output (`--filter=id,name,type,bitrate,alerts,path`)
    - Show extended artist info by providing artist name: `show artists [ARTIST]`

Changes:
  General
    - Reset now has it's own command `reset` and is replacing `monitor --reset`
    - Switched from GitHub to PyPi for update notifications
  Configuration
    - Bitrate now uses friendly names: `128`, `320`, `FLAC`
    - `smtp_recipient` is now `email`
    - `release_by_date` is now `by_release_date`
    - `release_max_days` is now `release_max_age`
  Refresh
    - When using time-machine, a second refresh is no longer necessary
  Show
    - `show --artists` has changed to `show artists`
    - `show --new-releases` has changed to `show releases`; default is 7

Fixes:
  General
    - Latest available version of deemon now stored in database to reduce unnecessary checks
    - Artists now sorted alphabetically regardless of case
    - Comma in artist name no longer treated as delimiter if name in quotes
  Config (command)
    - Artist config no longer set to global config when added
  Configuration
    - Options are now validated all at once, at initialization
    - New configuration options are now added to existing `config.json`
  Download
    - Fixes crash when downloading non-existence album ID
  Monitor
    - Artist names are now validated for more accurate results
  Refresh
    - All errors/warnings now printed together after refresh completes

deemon/core/config.py
import json
import logging
from copy import deepcopy
from pathlib import Path
from typing import Optional

from deemon.core.exceptions import ValueNotAllowed, UnknownValue, PropertyTypeMismatch
from deemon.utils import startup

logger = logging.getLogger(__name__)

ALLOWED_VALUES = {
    'bitrate': {1: "128", 3: "320", 9: "FLAC"},
    'alerts': [True, False],
    'record_type': ['all', 'album', 'ep', 'single'],
    'release_channel': ['stable', 'beta']
}

DEFAULT_CONFIG = {
    "check_update": 1,
    "debug_mode": False,
    "release_channel": "stable",
    "query_limit": 5,
    "smart_search": True,
    "rollback_view_limit": 10,
    "prompt_duplicates": False,
    "prompt_no_matches": True,
    "fast_api": True,
    "fast_api_threads": 25,
    "exclusions": {
        "enable_exclusions": True,
        "patterns": [],
        "keywords": [],
    },
    "new_releases": {
        "release_max_age": 90,
        "include_unofficial": False,
        "include_compilations": False,
        "include_featured_in": False,
    },
    "global": {
        "bitrate": "320",
        "alerts": False,
        "record_type": "all",
        "download_path": "",
        "email": ""
    },
    "deemix": {
        "path": "",
        "arl": "",
        "check_account_status": True,
        "halt_download_on_error": False,
    },
    "smtp_settings": {
        "server": "",
        "port": 465,
        "starttls": False,
        "username": "",
        "password": "",
        "from_addr": ""
    },
    "plex": {
        "base_url": "",
        "ssl_verify": True,
        "token": "",
        "library": ""
    }
}


class Config(object):
    _CONFIG_FILE: Optional[Path] = startup.get_config()
    _CONFIG: Optional[dict] = None

    def __init__(self):
        if not Config._CONFIG_FILE.exists():
            self.__create_default_config()

        if Config._CONFIG is None:
            with open(Config._CONFIG_FILE, 'r', encoding='utf8') as f:
                try:
                    Config._CONFIG = json.load(f)
                except json.decoder.JSONDecodeError as e:
                    logger.exception(f"An error occured while reading from config: {e}")
                    raise

            if self.validate() > 0:
                self.__write_modified_config()

            if not self.arl():
                logger.debug("Attempting to locate deemix's .arl file")
                arl_file = startup.get_appdata_root() / 'deemix' / '.arl'
                if Path(arl_file).is_file():
                    with open(arl_file) as f:
                        arl_from_file = f.readline().replace("\n", "")
                        self.set('arl', arl_from_file)
                        logger.debug("Successfully loaded ARL")

            if len(self.arl()) > 0 and len(self.arl()) != 192:
                logger.warning(f"   [!] Possible invalid ARL detected (length: {len(self.arl())}). ARL should "
                               "be 192 characters")

            # Set default for runtime settings
            self.set('profile_id', 1, validate=False)

    @staticmethod
    def __create_default_config():
        with open(Config._CONFIG_FILE, 'w', encoding='utf8') as f:
            json.dump(DEFAULT_CONFIG, f, indent=4)

    @staticmethod
    def __write_modified_config():
        with open(Config._CONFIG_FILE, 'w', encoding='utf8') as f:
            json.dump(Config._CONFIG, f, indent=4)

    @staticmethod
    def validate():
        modified = 0

        def process_config(dict1, dict2):
            """
            Process user configuration, applying values to a default config
            """
            for key, value in dict1.items():
                if key in dict2.keys():
                    if isinstance(dict1[key], dict):
                        process_config(dict1[key], dict2[key])
                    else:
                        dict2[key] = dict1[key]
            return dict2

        def find_position(d, key):
            for k, v in d.items():
                if isinstance(v, dict):
                    next = find_position(v, key)
                    if next:
                        return [k] + next
                elif k == key:
                    return [k]

        def update_config_layout(user_config, reference_config):
            """ Used to move existing values to new property names/locations """
            nonlocal modified

            if user_config.get('exclude'):
                user_config['exclusions'] = {
                    'enable_exclusions': True,
                    'patterns': user_config['exclude'],
                    'keywords': []
                }
                modified += 1

            if user_config.get('experimental'):
                if user_config['experimental'].get('allow_unofficial_releases'):
                    user_config['new_releases']['include_unofficial'] = True
                    modified += 1
                if user_config['experimental'].get('allow_compilations'):
                    user_config['new_releases']['include_compilations'] = True
                    modified += 1
                if user_config['experimental'].get('allow_featured_in'):
                    user_config['new_releases']['include_featured_in'] = True
                    modified += 1

            if user_config.get('new_releases'):
                if not user_config['new_releases'].get('by_release_date', True):
                    user_config['new_releases']['release_max_age'] = 0
                    modified += 1

            migration_map = [
                {'check_update': 'check_update'},
                {'plex_baseurl': 'base_url'},
                {'plex_token': 'token'},
                {'plex_library': 'library'},
                {'deemix_path': 'path'},
                {'arl': 'arl'},
                {'smtp_recipient': 'email'},
                {'smtp_server': 'server'},
                {'smtp_user': 'username'},
                {'smtp_pass': 'password'},
                {'smtp_port': 'port'},
                {'smtp_sender': 'from_addr'},
                {'bitrate': 'bitrate'},
                {'alerts': 'alerts'},
                {'record_type': 'record_type'},
                {'download_path': 'download_path'},
                {'release_max_days': 'release_max_age'},
                {'ranked_duplicates': 'prompt_duplicates'}
            ]
            for mlist in migration_map:

                for old, new in mlist.items():
                    user_config_tmp = deepcopy(user_config)
                    user_config_copy = user_config
                    if not user_config.get(old):
                        continue

                    old_position = find_position(user_config, old) or [old]
                    new_position = find_position(reference_config, new) or [new]

                    for i in old_position[:-1]:
                        user_config_tmp = user_config_tmp.setdefault(i, {})

                    for i in new_position[:-1]:
                        user_config_copy = user_config_copy.setdefault(i, {})

                    if user_config_tmp != user_config_copy:
                        logger.debug("Migrating " + ':'.join([str(x) for x in old_position]) + " -> " + ':'.join(
                            [str(x) for x in new_position]))
                        user_config_copy[new_position[-1]] = user_config_tmp[old_position[-1]]
                        modified += 1

            return user_config

        def test_values(dict1, dict2):
            nonlocal modified
            for key, value in dict1.items():
                if key in dict2.keys():
                    if isinstance(dict1[key], dict):
                        test_values(dict1[key], dict2[key])
                    else:
                        if key == "starttls" and value is True:
                            if dict1['port'] == 465:
                                logger.debug("STARTTLS enabled in config but port set to 465 (SSL); disabling STARTTLS")
                                dict1['starttls'] = False
                                modified += 1
                        if key in ALLOWED_VALUES:
                            if isinstance(ALLOWED_VALUES[key], dict):
                                if key == "bitrate" and value in ["1", "3", "9"]:
                                    if value == "1":
                                        dict1['bitrate'] = "128"
                                    if value == "3":
                                        dict1['bitrate'] = "320"
                                    if value == "9":
                                        dict1['bitrate'] = "FLAC"
                                    modified += 1
                                elif value in ALLOWED_VALUES[key].keys():
                                    dict1_tmp = dict1
                                    pos = find_position(dict1_tmp, key)
                                    for i in pos[:-1]:
                                        dict1_tmp = dict1.setdefault(i, {})
                                    dict1_tmp[key] = ALLOWED_VALUES[key][value]
                                    modified += 1
                                elif (
                                        value in ALLOWED_VALUES[key].values() or
                                        value.upper() in ALLOWED_VALUES[key].values()
                                ):
                                    continue
                                else:
                                    raise UnknownValue(
                                        f"Unknown value in config - '{key}': {value} (type: {type(value).__name__})")
                            elif not isinstance(dict1[key], type(dict2[key])):
                                if isinstance(dict2[key], bool):
                                    if dict1[key] == 1:
                                        dict1[key] = True
                                        modified += 1
                                    if dict1[key] == 0:
                                        dict1[key] = False
                                        modified += 1
                                else:
                                    raise UnknownValue(
                                        f"Unknown value in config - '{key}': {value} (type: {type(value).__name__})")
                            else:
                                if value in ALLOWED_VALUES[key]:
                                    continue
                                else:
                                    raise UnknownValue(
                                        f"Unknown value in config - '{key}': {value} (type: {type(value).__name__})")
                        elif not isinstance(dict1[key], type(dict2[key])):
                            if isinstance(dict2[key], bool):
                                if dict1[key] == 1:
                                    dict1[key] = True
                                    modified += 1
                                if dict1[key] == 0:
                                    dict1[key] = False
                                    modified += 1
                            else:
                                raise PropertyTypeMismatch(
                                    f"Invalid type in config - '{str(key)}' incorrectly set as {type(value).__name__}")
                        else:
                            pass

        def add_new_options(dict1, dict2):
            nonlocal modified
            for key, value in dict1.items():
                if key not in dict2.keys():
                    logger.debug(f"New option added to config: {key}")
                    dict2[key] = value
                    modified += 1
                elif isinstance(value, dict):
                    for k, v in value.items():
                        if k not in dict2[key].keys():
                            logger.debug("New option added to config: "
                                         f"{key}/{k}")
                            dict2[key][k] = v
                            modified += 1

        logger.debug("Loading configuration, please wait...")
        add_new_options(DEFAULT_CONFIG, Config._CONFIG)
        migrated_config = update_config_layout(Config._CONFIG, DEFAULT_CONFIG)
        Config._CONFIG = process_config(migrated_config, DEFAULT_CONFIG)
        test_values(Config._CONFIG, DEFAULT_CONFIG)
        return modified

    @staticmethod
    def get_config_file() -> Path:
        return Config._CONFIG_FILE

    @staticmethod
    def get_config() -> dict:
        return Config._CONFIG

    @staticmethod
    def plex_baseurl() -> str:
        return Config._CONFIG.get('plex').get('base_url')

    @staticmethod
    def plex_token() -> str:
        return Config._CONFIG.get('plex').get('token')

    @staticmethod
    def plex_library() -> str:
        return Config._CONFIG.get('plex').get('library')

    @staticmethod
    def download_path() -> str:
        return Config._CONFIG.get('global').get('download_path')

    @staticmethod
    def deemix_path() -> str:
        return Config._CONFIG.get('deemix').get('path')

    @staticmethod
    def arl() -> str:
        return Config._CONFIG.get('deemix').get('arl')

    @staticmethod
    def release_max_age() -> int:
        return Config._CONFIG.get('new_releases').get('release_max_age')

    @staticmethod
    def bitrate() -> str:
        return Config._CONFIG.get('global').get('bitrate')

    @staticmethod
    def alerts() -> bool:
        return Config._CONFIG.get('global').get('alerts')

    @staticmethod
    def record_type() -> str:
        return Config._CONFIG.get('global').get('record_type')

    @staticmethod
    def smtp_server() -> str:
        return Config._CONFIG.get('smtp_settings').get('server')

    @staticmethod
    def smtp_port() -> int:
        return Config._CONFIG.get('smtp_settings').get('port')

    @staticmethod
    def smtp_user() -> str:
        return Config._CONFIG.get('smtp_settings').get('username')

    @staticmethod
    def smtp_pass() -> str:
        return Config._CONFIG.get('smtp_settings').get('password')

    @staticmethod
    def smtp_sender() -> str:
        return Config._CONFIG.get('smtp_settings').get('from_addr')

    @staticmethod
    def smtp_recipient() -> list:
        return Config._CONFIG.get('global').get('email')

    @staticmethod
    def smtp_starttls() -> bool:
        return Config._CONFIG.get('smtp_settings').get('starttls')

    @staticmethod
    def check_update() -> int:
        return Config._CONFIG.get('check_update')

    @staticmethod
    def debug_mode() -> bool:
        return Config._CONFIG.get('debug_mode')

    @staticmethod
    def profile_id() -> int:
        return Config._CONFIG.get('profile_id')

    @staticmethod
    def update_available() -> int:
        return Config._CONFIG.get('update_available')

    @staticmethod
    def query_limit() -> int:
        return Config._CONFIG.get('query_limit')

    @staticmethod
    def prompt_duplicates() -> int:
        return Config._CONFIG.get('prompt_duplicates')

    @staticmethod
    def prompt_no_matches() -> bool:
        return Config._CONFIG.get('prompt_no_matches')

    @staticmethod
    def allowed_values(prop):
        return ALLOWED_VALUES.get(prop)

    @staticmethod
    def release_channel() -> str:
        return Config._CONFIG.get('release_channel')

    @staticmethod
    def rollback_view_limit() -> int:
        return Config._CONFIG.get('rollback_view_limit')

    @staticmethod
    def transaction_id() -> int:
        return Config._CONFIG.get('tid')
    
    @staticmethod
    def check_account_status() -> bool:
        return Config._CONFIG.get('deemix').get('check_account_status')
    
    @staticmethod
    def fast_api() -> bool:
        return Config._CONFIG['fast_api']

    @staticmethod
    def fast_api_threads() -> int:
        return Config._CONFIG['fast_api_threads']

    @staticmethod
    def allow_compilations() -> bool:
        return Config._CONFIG['new_releases']['include_compilations']

    @staticmethod
    def allow_featured_in() -> bool:
        return Config._CONFIG['new_releases']['include_featured_in']

    @staticmethod
    def allow_unofficial() -> bool:
        return Config._CONFIG['new_releases']['include_unofficial']

    @staticmethod
    def enable_exclusions() -> bool:
        return Config._CONFIG['exclusions']['enable_exclusions']

    @staticmethod
    def exclusion_keywords() -> list:
        if Config._CONFIG['exclusions']['enable_exclusions']:
            return [x.lower() for x in Config._CONFIG['exclusions']['keywords']]
        else:
            return []

    @staticmethod
    def exclusion_patterns() -> list:
        if Config._CONFIG['exclusions']['enable_exclusions']:
            return Config._CONFIG['exclusions']['patterns']
        else:
            return []

    @staticmethod
    def plex_ssl_verify() -> bool:
        return Config._CONFIG.get('plex').get('ssl_verify')

    @staticmethod
    def halt_download_on_error() -> bool:
        return Config._CONFIG.get('deemix').get('halt_download_on_error')

    @staticmethod
    def smart_search() -> bool:
        return Config._CONFIG.get('smart_search')


    @staticmethod
    def find_position(d, property):
        for k, v in d.items():
            if isinstance(v, dict):
                next = Config.find_position(v, property)
                if next:
                    return [k] + next
            elif k == property:
                return [k]

    @staticmethod
    def get(property):
        return Config._CONFIG.get(property)

    @staticmethod
    def set(property, value, validate=True):
        if not validate:
            Config._CONFIG[property] = value
        if Config._CONFIG.get(property):
            if property in ALLOWED_VALUES:
                if value.lower() == "true" or value == "1":
                    value = True
                elif value.lower() == "false" or value == "0":
                    value = ""
                if value in ALLOWED_VALUES[property]:
                    Config._CONFIG[property] = value
                    return
                raise ValueNotAllowed(f"Property {property} requires one of "
                                      f"{', '.join(ALLOWED_VALUES[property])}, not {value}.")

            if isinstance(value, type(Config._CONFIG[property])):
                Config._CONFIG[property] = value
                return
            else:
                raise PropertyTypeMismatch(f"Type mismatch while setting {property} "
                                           f"to {value} (type: {type(value).__name__})")

        else:
            property_path = Config.find_position(Config._CONFIG, property)
            tmpConfig = Config._CONFIG
            for k in property_path[:-1]:
                tmpConfig = tmpConfig.setdefault(k, {})
            if property in ALLOWED_VALUES:
                if isinstance(value, str):
                    if value.lower() == "true" or value == "1":
                        value = True
                    elif value.lower() == "false" or value == "0":
                        value = False
                if isinstance(ALLOWED_VALUES[property], dict):
                    if value in [str(x.lower()) for x in ALLOWED_VALUES[property].values()]:
                        tmpConfig[property_path[-1]] = value
                        return
                if value in ALLOWED_VALUES[property]:
                    tmpConfig[property_path[-1]] = value
                    return
                raise ValueNotAllowed(f"Value for {property} is invalid: {value} (type: {type(value).__name__})")

            if isinstance(value, type(tmpConfig[property])):
                tmpConfig[property] = value
                return
            else:
                raise PropertyTypeMismatch(f"Type mismatch while setting {property} "
                                           f"to {value} (type: {type(value).__name__})")


class LoadProfile(object):
    def __init__(self, profile: dict):
        logger.debug(f"Loaded config for profile {str(profile['id'])} ({str(profile['name'])})")
        # Rename keys to match config
        profile["profile_id"] = profile.pop("id")
        profile["base_url"] = profile.pop("plex_baseurl")
        profile["token"] = profile.pop("plex_token")
        profile["library"] = profile.pop("plex_library")

        # Append to config for debug output; Remove profile name from dict
        Config.set("profile_name", profile.pop("name"), validate=False)

        for key, value in profile.items():
            if value is None:
                continue
            Config.set(key, value)

        for key, value in Config.get_config().items():
            if isinstance(value, dict):
                for k, v in value.items():
                    if key in ['smtp_settings'] or k in ['arl', 'email', 'token']:
                        if v:
                            v = "********"
                    logger.debug(f"> {key}/{k}: {v}")
            else:
                logger.debug(f"> {key}: {value}")
